### ISSUE TITLE
[Bug]: LWC 3.7.0 and QGIS 3.34 raster - symbology

### ISSUE THREAD

**Issue by josemvm** [3liz/lizmap-web-client#4036]
### What is the bug?

qgis desktop 3.22.16
![imagem](https://github.com/3liz/lizmap-web-client/assets/10053874/cd21c64e-cc24-4c53-8b6f-30ea1933e392)
LWC 3.6.8
![imagem](https://github.com/3liz/lizmap-web-client/assets/10053874/c9e9b7dd-e50d-4eeb-b79a-7c10a56f5389)

===========
qgis desktop 3.28.13
![imagem](https://github.com/3liz/lizmap-web-client/assets/10053874/ac8c7e95-aef9-425e-9d69-8373984a0a2e)
LWC 3.7.0
![imagem](https://github.com/3liz/lizmap-web-client/assets/10053874/0ee2d0b5-333d-42b2-b7ab-33974964aabf)

### Steps to reproduce the issue

.

### Versions

**Versions :**

* Lizmap Web Client : 3.7.0
* Lizmap plugin : 4.1.1
* QGIS Desktop : 3.28.13
* QGIS Server : 3.28.13
* Py-QGIS-Server : not used
* QGIS Server plugin atlasprint : 3.3.2
* QGIS Server plugin lizmap_server : 2.8.1
* QGIS Server plugin wfsOutputExtension : 1.8.0

<details>
<summary>List of safeguards :</summary>
<br/>

* Mode : normal
* Allow parent folder : no
* Prevent other drive : no
* Prevent PG service : no
* Prevent PG Auth DB : no
* Force PG user&pass : no
* Prevent ECW : no
</details>

### Check Lizmap plugin

- [X] I have done the step just before in the Lizmap QGIS desktop plugin before opening this ticket. Otherwise, my ticket is not considered valid and might get closed.

### Operating system

Ubuntu 22.04

### Browsers

Firefox

### Browsers version

Firefox 120.0.1

**Comment by josemvm** [comment]
in case of linear interpolation, not for discrete or exact interpolation

**Comment by Gustry** [comment]
Similar to https://github.com/qgis/QGIS/issues/55651 ?

**Comment by josemvm** [comment]
no, only in case of linear interpolation type not for discrete or exact interpolation type

**Comment by Antoviscomi** [comment]
@josemvm in my LWC 3.7 the color ramp it's not rendered as in screenshot below
![Schermata del 2024-02-05 11-44-23](https://github.com/3liz/lizmap-web-client/assets/12639444/cbdaf0a9-eb0b-4b28-a907-63e9f2e14462)
il LWC 3.6.9 it seems compressed as in screenshot below
![Schermata del 2024-02-05 11-56-13](https://github.com/3liz/lizmap-web-client/assets/12639444/c11abd52-952c-4365-adb2-f13ac13a2ddb)

**Comment by josemvm** [comment]
@Antoviscomi thanks
maybe a qgis server bug...

**Comment by Gustry** [comment]
Lizmap is taking the JSON legend from QGIS Server, as you can see in the ticket I mentioned earlier.
You can open your "developer tools" (F12), check the request made to QGIS server `GetLegendGraphic` and check the content of the JSON response. It's in the webbrowser.

**Comment by josemvm** [comment]
@Gustry the url from lizmap doesn´t contain the `GetLegendGraphic`
myhost/index.php/lizmap/service?repository=ma&project=PIAAC-AC&**REQUEST=GetMap**&SERVICE=WMS&VERSION=1.3.0&FORMAT=image%2Fpng&STYLES=default&TRANSPARENT=true&LAYERS=lst_inv_ac&DPI=96&WIDTH=2080&HEIGHT=475&CRS=EPSG%3A3763&BBOX=-253278.93184485074%2C-179934.01850668347%2C297055.50215735065%2C-54256.68381868075&ts=1707138769681

i can replace the request for myhost/index.php/lizmap/service?repository=ma&project=PIAAC-AC&**REQUEST=GetLegendGraphic&FORMAT=image**&&SERVICE=WMS&VERSION=1.3.0&FORMAT=image%2Fpng&STYLES=default&TRANSPARENT=true&LAYERS=lst_inv_ac&DPI=96&WIDTH=2080&HEIGHT=475&CRS=EPSG%3A3763&BBOX=-253278.93184485074%2C-179934.01850668347%2C297055.50215735065%2C-54256.68381868075&ts=1707138769681

and get:
![imagem](https://github.com/3liz/lizmap-web-client/assets/10053874/bbd711eb-8db2-4dc5-bf52-cfd9e82b398d)

**Comment by Gustry** [comment]
This a GetLegendGraphic as a picture.
Now, Lizmap use the GetLegendGraphic as JSON, to be able to have a checkbox

![image](https://github.com/3liz/lizmap-web-client/assets/1609292/0f2cb4f9-f5e7-4cdc-8944-d45729875277)

**Comment by Antoviscomi** [comment]
@Gustry here in LWC 3.6.9 the request is:

http://myhost/webgis/index.php/lizmap/service?repository=tempor&project=ProvaFiles&SERVICE=WMS&VERSION=1.3.0&REQUEST=GetLegendGraphic&LAYER=orto2008&STYLE=predefinito&EXCEPTIONS=application%2Fvnd.ogc.se_inimage&FORMAT=image%2Fpng&TRANSPARENT=TRUE&WIDTH=150&LAYERFONTSIZE=0&ITEMFONTSIZE=9&SYMBOLSPACE=1&ICONLABELSPACE=2&DPI=96&RULELABEL=AUTO&LAYERSPACE=0&LAYERFONTBOLD=FALSE&LAYERTITLE=FALSE&SCALE=75000

in 3.7 there is a similar request in browser's 'developer tools'

**Comment by Gustry** [comment]
> in 3.7 there is not a similar request in browser's 'developer tools'

There is, check my screenshot, it's a POST request. There must be a `GetLegendGraphic` request ... it's the only way to ask QGIS server for the legend.

**Comment by Antoviscomi** [comment]
@Gustry there is not a request ` GetLegendGraphic ` on mine that's all
![Schermata del 2024-02-05 17-11-53](https://github.com/3liz/lizmap-web-client/assets/12639444/0fbdbf5e-1c0e-4544-b0ff-531a65a84f86)

**Comment by Antoviscomi** [comment]
only in Payload it is, and no response in LWC 3.7
`repository=tempor&project=ProvaFiles&SERVICE=WMS&REQUEST=GetLegendGraphic&VERSION=1.3.0&FORMAT=application%2Fjson&LAYER=edificicetraro_ctr%2Cterritori_boschi_foreste%2Corto2008&STYLES=predefinito%2Cpredefinito%2Cpredefinito`

**Comment by Gustry** [comment]
> there is not a request GetLegendGraphic on mine that's all

In your screenshot, it's the last one. Click one it, you will see in the `Request` tab, then open the "Response tab" and paste the content here

**Comment by Antoviscomi** [comment]
@Gustry 
this is the response

```
{
    "nodes": [
        {
            "symbols": [
                {
                    "title": "Band 1"
                },
                {
                    "title": ""
                }
            ],
            "title": "Temperatura Massima (\u00b0C)",
            "type": "layer",
            "name": "orto2008",
            "layerName": "Temperatura Massima (\u00b0C)"
        }
    ],
    "title": ""
}
```

**Comment by Gustry** [comment]
There isn't any `icon` key in the JSON.

This JSON is generated by QGIS Server, it's a graphical legend in JSON format.

This issue must be reported in QGIS. When QGIS will be fixed, it's high probability that the icon will be displayed in Lizmap Web Client then.

**Comment by Antoviscomi** [comment]
@Gustry why in LWC 3.6.9 it works? There is not the same request?

**Comment by Gustry** [comment]
* Before 3.7 : you can see your URL you have provided : `FORMAT=image/png`, QGIS server must give back to Lizmap Web Client a PNG file
* Now with 3.7, you can see `FORMAT=application/json`, QGIS server must give back a JSON representation of the legend.

So the bug is quite similar to https://github.com/qgis/QGIS/issues/55651 as I mentioned before. QGIS server is skipping some layers in the JSON representation.

**Comment by nboisteault** [comment]
@Gustry So we could use the `FORMAT=image/png` request for raster layers?

**Comment by josemvm** [comment]
> @Gustry So we could use the FORMAT=image/png request for raster layers?

@Gustry is it possible fix it with the workaround above? this is a little annoying... in production environment

thanks

**Comment by Gustry** [comment]
Yes @josemvm I know this issue, it's unfortunate and I understand...
But there is nothing I can do here, I'm doing the same answer as on https://github.com/3liz/lizmap-web-client/issues/4328#issuecomment-2110317484
Feel free to contact us.

**Comment by Gustry** [comment]
Given the work in https://github.com/3liz/qgis-lizmap-server-plugin/pull/77 we could notice that rendering the raster legend with QGIS < 3.34 is broken.

Supporting raster legend 3.7 will only work at least with QGIS 3.34.

@josemvm which version are you using ?
@Antoviscomi We can see your request that this is a QGIS Server < 3.34

**Comment by josemvm** [comment]
hi @Gustry i'am using qgis server 3.28

**Comment by josemvm** [comment]
> Supporting raster legend 3.7 will only work at least with QGIS 3.34

is the lwc 3.7 comaptible with qgis server 3.34?

**Comment by Antoviscomi** [comment]
> @Antoviscomi We can see your request that this is a QGIS Server < 3.34

@Gustry the above request was february 5 dated, now I'm on 3.36 version of QGisServer but I've not look at this issue anymore

**Comment by Gustry** [comment]
> ???

@josemvm We understand when there is a minimum of text and also when there is a single `?`. It looks like https://github.com/3liz/lizmap-web-client/issues/4377

>  this is a little annoying... in production environment

This is annoying as well to read, when we are spending so many "unpaid" time trying to make an opensource product.

**Comment by Antoviscomi** [comment]
> @Gustry the above request was february 5 dated, now I'm on 3.36 version of QGisServer but I've not look at this issue anymore, so I cannot tell It's yet here. I can test it asap if needed

@Gustry there is the request from LWC 3.7.7 and QGISServer 3.6.2
```
{
    "nodes": [
        {
            "symbols": [
                {
                    "title": "Band 1: Red (Red)"
                },
                {
                    "icon": "iVBORw0KGgoAAAANSUhEUgAAADMAAABgCAYAAACwu0pzAAAACXBIWXMAAA9hAAAPYQGoP6dpAAADLElEQVR4nO2cPWgUQRTHf7c3dwQLo6DBQjSiVooRLBRRRBPshdhoSvEDO4UoCqaIiCi2EgsLCytBEK2EIPgBAVExxESTKCqIH4Wa+EFy5G4t3orDGu8iu9mbPOYPw+7Nm9uZ\/\/3fezsL9zbXfKk1JAFeH+jNJfl+mjDFvDNrSQxTCIJ6ryE1mGJeD5mgmA9I0maIM0A\/8BN4B\/QAjZa9HQhj7ZZlXzSN\/Xt8kqxipgk4AgwAS4ErwEVgrzVmGGixPpenuU4z8DE6\/ytxZeVm+6zzDwiRztiYEJiocZ3JamNMsT4JYCvwJNa3DHgFlIA+oAt4ExtzFygAQ4jr3reN9UjN+4FWYJPVN4DEzTCwGDgJ3AHWA+NIrHUAjwGDuGcvsBl49PsippBtNjscLXQH8NLqfx41gFGE2BdgG3ATIXPVGn8c2AjswSaTYWo+ARwCtgMvaoydQhJAvsqYUtyeFZmziGvsROKgAQn4ycjejhB8i7hZF\/ANiRGALdFxCImZDkTd0\/YkphjMesw0AMei80Grv4z4P8A64AKwBHGvPqAN+BzZm4BuYAWiyCCwC7hnT5SFMhNArV\/sVNT+hetRqwpV2xlT0LRrrtNNc1agy81UPZwpU0YRGWWPzT5mnIQxgak9ao7A5HPVdtlzC7qUMV4ZN2Hymsh4N3MUPjW7Ch8zrsKnZlehzM28Mm5CFxnvZo7Cp2ZX4WPGVRjKU\/VeQ2owTGkio0oZVWS8m7kJE3pl3IQyMt7N3IShPN2fVOcmvJu5Cq+Mq1BGRpObhaqUUUVGk5t5ZRyFLmUqJUUbzVATmcqkIjJhqVLvNaQGHzOuQpcyyhKAIjKq3EyTMkGlVCFJm+k8SK3mJ+AHcAMpY0wVWSWATqRQtA14D1wGriEF26khKzc7iBSV9kefjyJF2WuAZ2lNkkUCaASWAw+tvhGkmryFNMlkoMz86DgW6\/9q2VJBFjEzHh0bY\/0LLFsqyEKZMaTSfAN\/3qywElHlaZoTZXXT7EFeCHIbyWbngQekGC8A5j\/uFUlwDliI1PfPQ14ZsTvtSXKja1cnelfTqoERZ8oJ\/a7ZVajaNf8C5uAfAthJht8AAAAASUVORK5CYII=",
                    "max": 255,
                    "min": 0,
                    "title": ""
                }
            ],
            "title": "550041_ORTO",
            "type": "layer",
            "name": "l_550041_ORTO",
            "layerName": "550041_ORTO"
        }
    ],
    "title": ""
}
```

and this is the reult on LWC 3.7.7 Canvas an TOC

![Schermata del 2024-05-16 16-12-43](https://github.com/3liz/lizmap-web-client/assets/12639444/70a9c5c6-b335-475b-a91a-282a470ef3d3)

the legend is now visible but however it seems a little bit small

**Comment by Gustry** [comment]
Thanks for checking @Antoviscomi

**Comment by Antoviscomi** [comment]
It's a pleasure @Gustry

**Comment by josemvm** [comment]
@Gustry  i'm sorry, you are right but i had no intention of hurting anyone or devaluing your work or project

**Comment by gioman** [comment]
> and this is the result on LWC 3.7.7 Canvas and TOC
> 
> ![Schermata del 2024-05-16 16-12-43](https://private-user-images.githubusercontent.com/12639444/331243675-70a9c5c6-b335-475b-a91a-282a470ef3d3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg1MzA3NzIsIm5iZiI6MTcxODUzMDQ3MiwicGF0aCI6Ii8xMjYzOTQ0NC8zMzEyNDM2NzUtNzBhOWM1YzYtYjMzNS00NzViLWE5MWEtMjgyYTQ3MGVmM2QzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE2VDA5MzQzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg1NmZiMzk5NWQ1ZGQ0Y2ZhMWQ1YzRiMDcwNTU5YWFlNDExMzUxYTY0ODM1MTViMTViN2M3NTgyNDgwYjA0YTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2Iq79-m_TpQQTyW5b8cMtvFn20l0MZkNCuQ0rzX9H3Y)
> 
> the legend is now visible but however it seems a little bit small


Doesn't this mean that the reported issue is fixed (meaning also that QGIS 3.34 is necessary?)?

**Comment by Gustry** [comment]
Using QGIS 3.34 is indeed better, but the image provided by QGIS server is too small.
Definitely do not use QGIS < 3.34 for raster symbology.

**Comment by Antoviscomi** [comment]
@Gustry @gioman in 3.36 version it's small also, so I think it's better to try to adjust the served image dimension before to render it in legend

**Comment by Antoviscomi** [comment]
The default values of height and width are 16 px, if its possible to change it only for raster legend, i .e. via bootstrap or via script, the image become as large as we need (this as workaround).
in the above screenshot I'd symply altered theese values in legend style to h = 75 px and W = 35 px 


![Schermata del 2024-06-17 10-33-25](https://github.com/3liz/lizmap-web-client/assets/12639444/fec3ef72-274c-4559-92ff-2de7c2499a7c)

**Comment by gioman** [comment]
> Using QGIS 3.34 is indeed better, but the image provided by QGIS server is too small.

@Gustry any reason why QGIS server returns (on purpose?) such small raster legend?

**Comment by Gustry** [comment]
I don't know. Maybe first developed for vector purpose/testing ?

According to the PR I made, the GetLegendGraphic JSON was working nicely on QGIS 3.16. Then with LTR 3.22 and 3.28, the JSON was broken, and then on 3.34, the JSON output is back, so no a lot of users on QGIS 3.22 and 3.28 it seems.

**Comment by gioman** [comment]
> I don't know. Maybe first developed for vector purpose/testing ?
> 
> According to the PR I made, the GetLegendGraphic JSON was working nicely on QGIS 3.16. Then with LTR 3.22 and 3.28, the JSON was broken, and then on 3.34, the JSON output is back, so no a lot of users on QGIS 3.22 and 3.28 it seems.

@Gustry not sure I follow: what would be the ideal fix here, patch QGIS Server to generate raster legends with a more reasonable size?

**Comment by josemvm** [comment]
ok, sorry...
>This a GetLegendGraphic as a picture.
>Now, Lizmap use the GetLegendGraphic as JSON, to be able to have a checkbox

**Comment by Gustry** [comment]
> i have tested on qgis server 3.28 and LWC 3.6.14-pre

@josemvm Before LWC 3.7, whatever the version of QGIS server, the format is PNG https://github.com/3liz/lizmap-web-client/issues/4036#issuecomment-1927391719

**Comment by josemvm** [comment]
>@josemvm Before LWC 3.7, whatever the version of QGIS server, the format is PNG https://github.com/3liz/lizmap-web-client/issues/4036#issuecomment-1927391719

@Gustry 3.22(.16)

**Comment by Gustry** [comment]
@josemvm "Whatever" means "regardless" ;-)
It doesn't matter the version of QGIS server used, only LWC version, requesting the legend in JSON format.

> @Gustry not sure I follow: what would be the ideal fix here, patch QGIS Server to generate raster legends with a more reasonable size?

@gioman Yes, being able to set the size of the picture would be better, or at least increase it by default, like the rendering in QGIS Desktop.

Doing the tip of @Antoviscomi works, but with text/fonts, this can lead to "transformed vertically" text

**Comment by gioman** [comment]
> @gioman Yes, being able to set the size of the picture would be better, or at least increase it by default, like the rendering in QGIS Desktop.

@Gustry so this is the ideal solution you are looking for, correct?

**Comment by gioman** [comment]
> Doing the tip of @Antoviscomi works, but with text/fonts, this can lead to "transformed vertically" text

I'm not sure I understand how the workaround works.

**Comment by Antoviscomi** [comment]
> I'm not sure I understand how the workaround works

@gioman I suggest to add a dimensional control in LWC that allow to manage the height/width of png legend served by QGIS server (now the default dimension readed by LWC is 16x16px, that seems very small) so a control (hardcoded in LWC css i.e.) should permit to well adapts these dimensions

**Comment by josemvm** [comment]
>@josemvm Before LWC 3.7, whatever the version of QGIS server, the format is PNG https://github.com/3liz/lizmap-web-client/>issues/4036#issuecomment-1927391719
@Gustry i misinterpreted your affirmation
thanks

**Comment by josemvm** [comment]
may an option in the lizmap plugin to _GetLegendGraphic_ as a picture/png, only for lwc 3.7 (?)

**Comment by Gustry** [comment]
Thanks @Antoviscomi for the correction, indeed, way better !

@josemvm yes, we have talked about that internally. It might be needed maybe for some rendering engine, for like proportional circle. Feel free to contact us.

**Comment by Antoviscomi** [comment]
@Gustry @gioman @josemvm 

![Senzanome](https://github.com/3liz/lizmap-web-client/assets/12639444/bdb5d91f-8406-4023-bbbe-829d2d2a72dd)
setting 

`heigth: 'auto; `
in map.css at 

```
lizmap-treeview img.legend{
    height: 16px;
    width: 16px;
}
```
the png raster legend it seems to works as expetted

**Comment by Gustry** [comment]
Thanks @Antoviscomi for the fix
Fixed in next 3.7 and 3.8
